src/passes/OptimizeInstructions.cpp

```diff
@@ -3549,6 +3558,38 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // Bitwise AND of a value with bits in [0, n) and a constant with no bits in
+  // [0, n) always yields 0. Replace with zero.
+  Expression* optimizeAndNoOverlappingBits(Binary* curr) {
+    assert(curr->op == AndInt32 || curr->op == AndInt64);
+
+    using namespace Abstract;
+    using namespace Match;
+
+    auto* left = curr->left;
+    auto* right = curr->right;
+
+    // Check left's max bits and right is constant.
+    auto leftMaxBits = Bits::getMaxBits(left, this);
+    uint64_t maskLeft;
+    if (leftMaxBits == 64) {
```

Comment:
This would be 32 for a 32-bit operation.
```suggestion
    if (leftMaxBits == left->type.getByteSize() * 8) {
```

Reply:
For a binaryOp `AndInt32` or `AndInt64`, is it proper that the type of the left operand is `None` or `unreachable`? It does not make sense.

Reply:
None is not valid IR, but unreachable is possible, like
```
(i32.add
  (i32.const 42)
  (unreachable)
)
```
An unreachable like that can appear anywhere in the IR. The dead code elimination pass will remove the add in this case (since it is never reached), but that IR is still valid.

Reply:
Thanks for teaching, and I've updated by adding the check `isNumber`, which leaves `None` and `unreachable` to DCE.